Switch to untyped AST in the TinyLang.Field.Parser

jakzale opened the pull request:
We introduce a Raw (untyped) AST for `TinyLang.Field.Parser`

- fix some `haddock` errors, so that `tiny-lang` is searchable by `stack hoogle`,
- add `TinyLang.Field.Raw.Parser` for producing untyped AST (#47),
- refactor existential qualification tag to `TinyLang.Field.Existential`,
- rafactor `Uni`, `KnownUni`, `UniVal`, and `SomeUniVal` to `TinyLang.Field.UniVal`,
- generalise `Data.Field.TextField` typeclass to all parsers working on `String`,
- generalise `TinyLang.ParseUtils` to all parsers working on `String`,
- add `parseString` to `TinyLang.ParseUtils`.

Commit:
HOTFIX for building Haddock

Commit:
Start working on untyped parser

Commit:
Reorder syntactic classes a bit

Commit:
WIP Implementation

Commit:
Finish pIdentifier

Commit:
Add support for literal

Commit:
Start testing untyped parser

For the time being I am disabling all other tests

effectfully commented on field/TinyLang/Field/Generator.hs:
```diff
@@ -317,19 +317,17 @@ boundedArbitraryExprI
+  reduce the risk of this, or supply a separate list of variables
+  which we're certain will only contain integer values.  This Note
+  also applies to the @size <= 1@ case above.
+-}
```
That comment only applies to the `, (0, EVar <$> chooseUniVar vars)` line, so it shouldn't be moved away from it (but feel free to move it before the line rather than after it).

jakzale replied:
`haddock` version specified in `lts-13.26` breaks for that syntax.

Will check in spare time, if it was fixed in later version of `haddock`.

effectfully replied:
Don't bother, just remove `^` to make it a regular comment.

You found that out by running `haddock`? That's great.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+  | CInt   Integer
+  | CVec   [Bool]
+
+data Var = Var Identifier
```
We don't need to define another `Var` here. Instead, `TinyLang.Var.Var` should be used.

jakzale replied:
I was not sure about it.  `TinyLang.Var.Var` contains `Unique` field, which I believe is used to tell variables apart.

Since I am not checking for scopes at this stage, how should I generate values of `Unique`?

effectfully replied:
In

```haskell
parseExpr
    :: forall f m. (MonadSupply m, TextField f)
    => String -> m (Either String (SomeUniExpr f))
parseExpr = fmap _scopedValue . parseScopedExpr
```

the `MonadSupply` constraint is responsible for generation of fresh `Unique`s.

In `common/TinyLang/ParseUtils.hs` we have these definitions:

```haskell
type Parser = ParsecT Void String (StateT Scope Supply)

-- | 'Scope' maps names onto 'Var's.
type Scope = Map String Var

data Scoped a = Scoped
    { _scopedScope :: Map String Var
    , _scopedValue :: a
    } deriving (Functor, Foldable, Traversable)

-- | Look up a variable name. If we've already seen it, return the corresponding Var;
-- otherwise, increase the Unique counter and use it to construct a new Var.
makeVar :: String -> Parser Var
makeVar name = do
    vars <- get
    case Map.lookup name vars of
        Just var -> pure var
        Nothing  -> do
            var <- freshVar name
            put $ Map.insert name var vars
            pure var
```

There are two options: either we keep these definitions or do not assign any `Unique`s to variables during parsing and instead do that in a separate pass, which forces us to parameterize the untyped AST by `var`, which gets instantiated to `String` during parsing and to `TinyLang.Var.Var` after the separate pass.

I'm slightly in favor of the latter option, because it also allows to instantiate `var` with a type of De Bruijn indices, which is handy for tests, compression, etc (we use that trick in Plutus). But the former option is fine too as it's fewer changes to make and transitioning from the former to the latter is a straightforward thing to do, so we can do that once we actually need it.

jakzale replied:
Went with parameterising the AST by the type of the variable.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+data Const
+  = CBool  Bool
+  | CInt   Integer
+  | CVec   [Bool]
```
`TinyLang.Field.Core.SomeUniVal` should be used here. No point to introduce another definition.
Except raw terms shouldn't depend on typed terms, so `TinyLang.Field.Core.SomeUniVal` should be extracted somewhere common and both the typed and raw ASTs would depend on it.

jakzale replied:
Do we want to have the untyped language and typed language depend on some common types?

effectfully replied:
Yes.

jakzale replied:
Then I guess I should refactor out:
- `Uni`, and
- `SomeUniVal`.

As `SomeUniVal` depends on `Uni`.

effectfully replied:
Yes.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+
+data Var = Var Identifier
+
+data Expr
```
Should have the `f` parameter. `Integer` is not a thing at the moment (it may become later), currently we only have an abstract field.

jakzale replied:
Should the `Raw` AST be parameterised by `f`, or should the `f` be existentially quatified within the type?

effectfully replied:
Parameterized.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+-- Identifier Character
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c =
+  or $ map (\f -> f c) [isLower , isDigit, (=='_'), (=='\'')]
```
`or . map f` is better known as `any f`.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer      as L
+import           Data.Void
```
We have `TinyLang.Prelude` that reexports a lot of stuff including `Data.Void`. It's much less annoying to develop things when you don't need to import `Data.Maybe`, `Data.List`, `Control.Monad`, etc all the time, so I'd suggest to replace `Data.Void` with `TinyLang.Prelude`, which also reexports `Set`, so you don't need to write the ugly looking `S.Set`.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+type Identifier = String
+
+data Const
+  = CBool  Bool
```
We use 4 spaces for indentation instead of 2 in Plutus, so I decided to keep that formatting rule in this project as well. Here's what I have in my `emacs` config:

```
(setq tab-stop-list (number-sequence 4 200 4))
(custom-set-variables
 '(haskell-indentation-layout-offset 4)
 '(haskell-indentation-left-offset 4)
 '(haskell-indentation-starter-offset 4)
 '(haskell-indentation-where-post-offset 4)
 '(haskell-indentation-where-pre-offset 4)
 '(tab-width 4)
 ...
)

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+
+identifierChar :: (MonadParsec e s m, Token s ~ Char) => m (Token s)
+identifierChar = satisfy isIdentifierChar <?> "identifier character [a-z0-9_']"
+{-# INLINE identifierChar #-}
```
Adding `INLINE` pragmas should be done together with benchmarking (I've seen cases where my code became slower after I sprinkled `INLINE` pragmas over it where I thought they would make sense), so don't bother adding this syntactic noise right now.

jakzale replied:
:+1:

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
```
I thought `Text.Megaparsec` reexports the functions that are hidden here? If you remove `many` and `some`, will there be errors? What are they?

jakzale replied:
Ah, for some reason I thought that `between` was not in `Text.Megaparsec`.  Must have checked the wrong version of `megaparsec`.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+pIdentifier =
+  lexeme $ do
+  prefix     <- option "" (string "?" <|> string "#")
+  identifier <- (:) <$> lowerChar <*> many identifierChar
```
`many` has terrible performance when it comes to parsing. Instead of `many (satisfy p)` (`identifierChar` expands to `satisfy`) use `takeWhileP`. [Proof](https://markkarpov.com/post/megaparsec-more-speed-more-power.html#there-is-hope).

We will eventually move from `String` to `Text`, so this is another reason to use `takeWhileP`.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+  lexeme $
+  between
+  (symbol "{")
+  (symbol "}")
```
Hard to read without any indentation.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,258 @@
+  between
+  (symbol "{")
+  (symbol "}")
+  ((pBoolLiteral `sepBy` (symbol ",")) >>= (pure . concat))
```
``concat <$> (pBoolLiteral `sepBy` symbol ",")``?

jakzale replied:
Yep, I am still getting used to applicative-style programming.

Commit:
Add .stylish-haskell.yaml

Commit:
Fix `or . map` to `any`

Commit:
Start adding types to the parser

Commit:
Fix whitespace

Commit:
Remove redundant import to Control.Applicative.Combinators

Commit:
Switch to takeWhileP

Commit:
Fix indentation to 4 spaces

Edit the code to be slightly more readable.

Commit:
Use TinyLang.Prelude

Commit:
Revert "HOTFIX for building Haddock"

This reverts commit 2fe47ad77436a3e06846d72f1902f1cbf59bb9f8.

Commit:
FIX Wrong Haddock comments

effectfully commented:
Forgot to add: it's fine to keep the raw AST and the parser in the same module during development, but before merging the PR we'll need to separate the two ASTs and their infrastructure to different folders, so that we'll have the following file structure:

- `TinyLang.Field.Typed.Core`
- `TinyLang.Field.Typed.Evaluator`
- `TinyLang.Field.Raw.Core`
- `TinyLang.Field.Raw.Parser`
- etc

We'll drop the `Field` prefix once we get rid of the `boolean` language.

effectfully commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,285 @@
+
+pIntLiteral :: Parser Integer
+pIntLiteral =
+    lexeme $ fmap read (some digitChar)
```
`some` is as inefficient as `many`.

Commit:
Playing code golf with fmap and <$>

Commit:
WIP Operations

Commit:
Finish adding operators

Commit:
Reorder operator precedence a bit

jakzale commented on field/TinyLang/Field/UParser.hs:
```diff
@@ -0,0 +1,348 @@
+      ]
+    , [ unary  (keyword "unpack") $ Unp
+      ]
+    ]
```
@effectfully, what is the operator precedence for this language?

The original spec mentions the following precedences:
- `not > xor > and > or`, and
- `{neg,inv} > {*,/} > {+,-}`

but it does not mention operator precedence for the following
- `neq0`
- `==, <=, <, >=, >`
- `unpack`
- `expr [ expr ]`

So far I assumed that:
- `not` > everything
- arithmetic > `neq0`,
- arithmetic > boolean, and 
- everything  > unpack.

effectfully replied:
All prefix unary operators (`not`, `neq0`, `neg`, `inv`, `unpack`) should have the same precedence that is greater than the precedences of all the binary operators. And both `unOp1 (unOp2 x)` and `unOp1 (x binOp2 y)` always require parentheses. `unOp1 x binOp unOp2 y` never requires parentheses.

`expr [ expr ]` is a postfix operator. It should have precedence that is less than the one of unary operators, but still greater than the precedences of all the binary operators. So that `unOp x [ expr ]` never requires parentheses, but `(x binOp y) [ expr ]` always requires them.

Binary operators are handled as per usual like you've assumed.

jakzale replied:
What is the difference between:
- `unOp1 (x binOp2 y)`, and
- `unOp1 x binOp2 y`
?

effectfully replied:
Sorry for being unclear.

- `unOp1 (x binOp2 y)` parses as `unOp1 (x binOp2 y)`
- `unOp1 x binOp2 y` parses as `(unOp1 x) binOp2 y`

effectfully replied:
Err,

> Implements the order specified by @effectfully, however it does permit
the following syntax `a + b [c]` where `[c]` is binding as a postfix operator to `b`.

by my "`(x binOp y) [ expr ]` always requires parentheses" I mean that if the user wants an expression to parse like that, then parentheses are required. If the user wants to parse an expression like `x binOp (y [ expr ])`, then parentheses are not required and both of

- `x binOp y [ expr ]`
- `x binOp (y [ expr ])`

are fine. This is what I wanted to say by

> expr [ expr ] is a postfix operator. It should have precedence that is <...> greater than the precedences of all the binary operators. 

And this same logic applies to other precedence-related rules that I mentioned.

So what you implement is exactly what is desired, there is no "however" there.

Commit:
Refactor some to takeWhile1P

Commit:
Index Expr and Statement by f

Commit:
Potential fix for operator precedence.

Implements the order specified by @effectfully, however it does permit
the following syntax

```
a + b [c]
```

where `[c]` is binding as a postfix operator to `b`.

Commit:
Move UParser to Raw.Parser

Commit:
WIP Parser

Commit:
Remove duplicated rule for `expr [ expr ]`

Commit:
Remove stylish haskell

jakzale commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,360 @@
+    "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+    (statement (";" statement)*)?
```
@effectfully, I have a question about parsing `statements`:  should the grammar permit empty statements or empty expressions?

First of all, I think you already mentioned that the grammar is slightly akward, with statements being the other way round, and currently, the smallest program with a statement is
```
assert T; T
```
For for loops the original grammar from `Parser.hs` permits `statements` to be empty, we should be able to write
```
for x = 0 to 10 do end; T
```
but the AST does not permit the above.

Theoretically, we can permit empty loops.  Should I edit the grammar to permit them?

Also, neither grammar nor ast permit the following program:
```
assert T;
```

effectfully replied:
> I have a question about parsing statements: should the grammar permit empty statements or empty expressions?

We don't want any empty expressions. We may want empty statements in the raw AST just for the ease of targeting and optimizing it (fewer invariants to preserve). We'll then need to remove them when mapping the raw AST to the typed one.

Permitting empty loops may be convenient for program generation (again, fewer invariants to worry about).

So yeah, I think should we allow empty loops. Please edit the grammar.

Oh, I totally forgot. Could you also add `for` loops to the AST and parse them directly instead of unrolling them on the fly like it was done previously? Like

```haskell
data Statement f
    = <...>
    | EFor Var (Expr f) (Expr f) [Statement f]
```

(a loop variable, a lower bound, an upper bound, a list of statements). Feel free to either do that here or in a separate PR depending on what is more convenient.

> Also, neither grammar nor ast permit the following program:
>
> `assert T;`

Yeah, that is something that we'll fix later.

Commit:
Start working on golden testing

- add # as line comment
- input files as test/Field/raw/*.field
- golden files as test/Field/raw/*.golden
- at the moment will load all files automatically
- add Glob to list directory for all files

Commit:
Add for loop and some tests

Commit:
Fix for loops

Commit:
Clean up the Raw.Parser a bit

Commit:
Start refactoring Uni out

Commit:
Add more raw parser tests

Commit:
Parameterise Raw AST by the type of Variable

Commit:
WIP SomeUniVar

effectfully commented on field/TinyLang/Field/Existential.hs:
```diff
@@ -0,0 +1,23 @@
+module TinyLang.Field.Existential
+    ( Some (..)
```
Great that you've extracted this to a separate module. We'll eventually need to make a library out of the [much nicer stuff](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/src/Language/PlutusCore/Universe/Core.hs) that I've implemented in Plutus recently and you it here as well.

Commit:
Generalise ParseUtils

Commit:
Generalise parseField and parseFieldDefault

Commit:
Fix the TextField instance for AField

Commit:
Switch to Rational

Commit:
Disable debugging

Commit:
Finish switch to rational

- Decided to use `EConst` instead of `EVal`.

jakzale commented:
@effectfully, I think this is getting pretty much done.

I will clean it up tomorrow, and then I will mark it ready for review.

Commit:
Remove comments from .field files

Commit:
Add utility functions to ParseUtils

- Exposed top
- Added brackets

Commit:
Add TinyLang.Field.Raw.Core

Commit:
Cleanup TinyLang.Field.Raw.Parser

- Use TinyLang.Field.Raw.Core
- Use TinyLang.ParseUtils

Commit:
Re-enable old test-suite

Commit:
Clean whitespace a bit

Commit:
Rename TinyLang.Field.Uni to TinyLang.Field.UniVal

Commit:
Add explicit exports to TinyLang.Field.UniVal

effectfully commented on common/TinyLang/ParseUtils.hs:
```diff
@@ -56,25 +58,35 @@ makeVar name = do
 
-signedDecimal :: Integral a => Parser a
+brackets :: (MonadParsec e s m, Token s ~ Char, Tokens s ~ [Char]) => m a -> m a
+-- brackets :: Parser a -> Parser a
```
Drop the comments, they outdate quickly.

effectfully commented on field/TinyLang/Field/Raw/Core.hs:
```diff
@@ -0,0 +1,71 @@
+type RawExpr = Expr Var
+type RawStatement = Statement Var
+
+{-| = Type Instances
```
`type instance` is what you write for a clause of an open type family. I think it's common to call the things below just "instances" (or "class instances" or "type class instances").

effectfully commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,339 @@
+We are using the following convention
+
+* /tokens/ are enclosed in parentheses @""@.
+* /regular expressions/ are enclosed in square brackets @[]@.
```
I think e.g. `[0-9]+` is a regular expression as a whole and `[0-9]` is just a part of it. I'd probably just drop this line.

effectfully commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,339 @@
+A simple parser for language involving booleans, field elements, and
+vectors.
+
+Given that the language is simple enough we are mixieng lexical
```
Typo: "mixieng".

effectfully commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,339 @@
+if-then-else has to be parenthesised unless it's at the very top.
+
+Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+Things like "neg inv 5" are illegal: use parentheses.
```
The operator precedence needs to be either fixed (to include all the possible case that we talked about before: `unOp x binOp y`, etc) or just dropped.

effectfully commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,339 @@
+pTop = top pExpr
+
+{-| == Helper functions
+-}
```
I don't think we should export them from here. `parseRational` is too specific and unsafe, if the caller needs that, they can implement it themselves. Let's instead export a function that receives a parser and a string and outputs an `Either String` (or `Either Text` depending on what is more convenient). Without hardcoding the field to `Rational`.

jakzale replied:
Ok,
moving `parseRational` to `test/Field/Raw/Textual`.

Added `TinyLang.ParseUtils.parseString` which evaluates to `Either String a`.

effectfully commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,339 @@
+      $ runParser (pTop @Rational) fileName str
+
+p :: String -> IO ()
+p s = putStrLn $ parseRational "" s
```
Let's not ever export one-letter names (or even create them on the top level, unless it's a quasi-quote or some other very special case).

effectfully commented on library/Data/Field.hs:
```diff
@@ -98,14 +98,15 @@ makeDiv d x y = do
 fromDivided = fromMaybe $ throw DivideByZero
 
-parseFieldDefault :: Field f => Parser f
+-- parseFieldDefault :: Field f => Parser f
```
Drop the comment.

effectfully commented on test/Field/Raw/golden/operator-precedence-00.field:
```diff
@@ -0,0 +1,12 @@
+	- c * d
+    >=
+	  e / f
+	+ g / h
```
It would be really nice to have another file alongside these two where we pretty-print the result of parsing to see how things get formatted. But that can be a separate PR.

jakzale replied:
I will add a pretty-printing golden testing pass as part of #30.

effectfully commented on test/Field/Raw/Textual.hs:
```diff
@@ -0,0 +1,49 @@
+gen_test_parsing :: IO TestTree
+gen_test_parsing =
+    do
+        files <- testFiles
```
So annoying that you can't do `IO` in `TestTree`.

jakzale replied:
Yes.  I am afraid that was intentional:
https://stackoverflow.com/a/33046723

effectfully commented on test/Main.hs:
```diff
@@ -15,4 +16,12 @@ test_all =
 main :: IO ()
-main = defaultMain test_all
+main =
+    do  test_rawParsing <- Raw.gen_test_parsing
```
Why not alter `test_all` above (and change its type to `IO TestTree`) instead of inlining everything here?

effectfully commented on field/TinyLang/Field/Raw/Core.hs:
```diff
@@ -0,0 +1,71 @@
+    | Neg
+    | Inv
+    | Unp
+
```
I'd write `deriving (Show)` here instead of having a standalone deriving declaration at the bottom as the latter results in a lot of boilerplate when there are many instances to derive.

Applies to other data types as well.

effectfully reviewed (approved):
LGTM. Comments are mostly nitpicking.

effectfully commented on field/TinyLang/Field/Raw/Parser.hs:
```diff
@@ -0,0 +1,339 @@
+import           TinyLang.Field.Raw.Core
+import           TinyLang.Field.Existential
+import           TinyLang.Field.UniVal
+import           Data.Set ( fromList
```
We have a convention that imports from external libraries and imports of internal modules are separated by a newline. The latter go first and the former go next (after a newline). This makes it easier to read imports.

Commit:
Remove comments

Commit:
Move instances next to data types.

Commit:
Move test generation to test_all

Commit:
Fix description of order precedence

Commit:
Clean up imports for Raw Parser

Commit:
Fix haddock comments

Commit:
Add TinyLang.ParseUtils.parseString

Commit:
Move parseRational to Field.Raw.Textual

jakzale commented:
@effectfully, incorporated your feedback.

Do you prefer to merge or squash-merge?

effectfully commented:
> Do you prefer to merge or squash-merge?

The latter unless some particular care was taken to organize commits to be well-structured, self-contained, etc.
